Fix: Pickup item cutscene logic fix

aMannus opened the pull request:
Resolves bug introduced by https://github.com/HarbourMasters/Shipwright/pull/1692 and https://github.com/HarbourMasters/Shipwright/issues/1734.

Turns out the sticks from the deku baba's aren't item00, and I missed the check that was there before my own changes. This meant that in rando, picking up sticks from a deku baba was showing the "get item" animation every time you picked up a stick from them. This fixes that.

On top of that, there was a check in `Item_CheckObtainability()` for keys and compasses, but those IDs aligned with slingshot seeds, so it was returning ITEM_NONE even when the player already picked up the seeds before. So that meant the cutscene would always play in rando.

I've looked into if removing these from `Item_CheckObtainability()`, but this function is only used in vanilla functions for scrubs and shops (which don't check for keys/maps/compasses), and some functions in z_player.c. The functions in z_player.c are either for the pick-up item cutscene logic which this PR already addresses, and seemingly some functions that decide if a chest should give out a blue rupee.

That said, I still can't say with 100% certainty I understand what's happening in the other functions in z_player.c or why they are there or if they can cause issues. I'm still looking into that, but initial testing with a rando-specific key in a chest seems to work fine.

aMannus committed:
Fix: Fix for the pickup CS logic fix (fixed)

aMannus committed:
Tiny cleanup

briaguya0 commented:
Could https://github.com/HarbourMasters/Shipwright/issues/1734 be related? Is there more to fix here?

aMannus commented:
> Could #1734 be related? Is there more to fix here?

Yeah, apparently so. I'll see if I can get to this as well soon.

aMannus commented:
Actually found some more things wrong with the logic here, so I'll look into fixing everything I've found first.

aMannus committed:
More fixes + clarified comments

aMannus commented:
There was quite a bit more to this. I'll edit the OP to reflect the new changes. This is good to go again and fixes https://github.com/HarbourMasters/Shipwright/issues/1734 now.

EDIT: Not good to go. I'm just not confident enough that this doesn't break other things just yet.

aMannus commented:
After further inspection, it looks like the rando-specific bottle items are also not checked for at all in Item_CheckObtainability(). On top of that, RG_BOTTLE_WITH_POE has the same ID as ITEM_SHIELD_DEKU (which IS checked for in the function), so I'm guessing if removing the keys/compasses/maps in this function would lead to problems, it should've already happened with bottled poes before as well.

I'm not super happy with not completely understanding the functions in z_player.c that use Item_CheckObtainability(), but I can't seem to break it either when removing the entries. And everything I do understand about this seems to point to this change being fine. So I'm still going to call this good to go.

dcvz commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -6246,20 +6246,19 @@ s32 func_8083E5A8(Player* this, GlobalContext* globalCtx) {
+                uint8_t skipItemCutsceneRando = gSaveContext.n64ddFlag && Item_CheckObtainability(giEntry.itemId) != ITEM_NONE && isDropToSkip;
+
+                // Show cutscene when picking up a item that the player doesn't own yet or hasn't picked up before yet.
                 if ((globalCtx->sceneNum == SCENE_BOWLING || Item_CheckObtainability(giEntry.itemId) == ITEM_NONE || gSaveContext.n64ddFlag) && !skipItemCutscene && !skipItemCutsceneRando) {
```
Could we also extract this first or checks with a comment? Similar to how you've done for the others?

aMannus replied:
You mean extract `(globalCtx->sceneNum == SCENE_BOWLING || Item_CheckObtainability(giEntry.itemId) == ITEM_NONE || gSaveContext.n64ddFlag)` into a uint8_t as well?

aMannus replied:
I'm going to assume it was that, that's done now.

aMannus committed:
Extract more logic into bool, clarified comments

aMannus commented:
Extracted more logic into its own bool and retested everything afterwards again. Also added more comments. By now the comment block is quite massive for this piece of code, but I think it's neccesary that whenever someone else has to work with this in the future doesn't run into the same problems as I did while trying to fix this code. It's become quite a spiderweb and I'm sure the logic could be cleaned up somehow, but I think how it is currently conveys the intention the best.

What a ride this has been for a seemingly simple fix lol.

dcvz reviewed (approved):
This is much more understandable. It's bulky but let's consider it as what it is.. documentation. :shipit:
